Framework/PythonInterface/plugins/algorithms/MatchSpectra.py

```diff
@@ -0,0 +1,220 @@
+        '''Convert to a Workspace2D despite what the algorithm is named'''
+        outputWS = ConvertToMatrixWorkspace(InputWorkspace=self.getPropertyValue('InputWorkspace'),
+                                            OutputWorkspace=self.getPropertyValue("OutputWorkspace"))
+        return outputWS
+
+    def __generateIndices(self, spectrumNum, reference, testing, binBoundaries):
+        '''Generates the indices for slicing by comparing x-axes
+
+        A note about implementation: If numpy.searchsorted fails to find the
+        value, it returns the last index of the array.
+        '''
+        BAD_RESULT = (False, (0, 0), (0, 0))
+        # find the lower bounds
+        refLower = 0
+        tstLower = 0
+        if reference[0] == testing[0]:
+            pass  # values are already set
+        elif reference[0] < testing[0]:
+            refLower = np.searchsorted(reference, testing[0])
+            if refLower == reference.size:
+                msg = 'Falied to find {} in reference spectrum x-axis (spectrum={})'.format(testing[0], spectrumNum)
+                self.log().notice(msg)
+                return BAD_RESULT
+        else:
+            tstLower = np.searchsorted(testing, reference[0])
+            if tstLower == testing.size:
+                msg = 'Falied to find {} in the x-axis of the spectrum being matched (spectrum={})'.format(reference[0],
+                                                                                                           spectrumNum)
+                self.log().notice(msg)
+                return BAD_RESULT
+
+        # find the upper bounds
+        refUpper = reference.size-1
+        tstUpper = testing.size-1
+        if binBoundaries:
+            refUpper -= 1
+            tstUpper -= 1
+        if reference[refUpper] == testing[tstUpper]:
+            pass  # values are already set
+        elif reference[refUpper] < testing[tstUpper]:
+            tstUpper = np.searchsorted(testing, reference[refUpper])
+            if reference[refUpper] != testing[tstUpper]:
+                msg = 'Falied to find {} in the x-axis of the spectrum being matched (spectrum={})'.format(reference[-1],
+                                                                                                           spectrumNum)
+                self.log().notice(msg)
+                return BAD_RESULT
+        else:
+            refUpper = np.searchsorted(reference, testing[tstUpper])
+            if reference[refUpper] != testing[tstUpper]:
+                msg = 'Falied to find {} in reference spectrum x-axis (spectrum={})'.format(testing[-1], spectrumNum)
+                self.log().notice(msg)
+                return BAD_RESULT
+
+        if (reference[refLower:refUpper]).size != (testing[tstLower:tstUpper]).size:
+            self.log().notice(msg)
+            return BAD_RESULT
+
+        return (True, (refLower, refUpper), (tstLower, tstUpper))
+
+    def __residual(self, X, Y1, Y2):
```

Comment:
You are not calculating chi^2 correctly. You need to add errors.

Reply:
It is the unweighted chi^2 and the documentation has been updated accordingly.